base_layer/core/src/consensus/consensus_manager.rs

```diff
@@ -20,36 +20,144 @@
 
-use crate::consensus::emission::EmissionSchedule;
+use crate::{
+    chain_storage::{BlockchainBackend, ChainStorageError},
+    consensus::emission::EmissionSchedule,
+    proof_of_work::{DiffAdjManager, DiffAdjManagerError, Difficulty, DifficultyAdjustmentError, PowAlgorithm},
+};
+use derive_error::Error;
+use std::sync::{Arc, RwLock};
+use tari_utilities::epoch_time::EpochTime;
 
-/// This is the used to control all consensus values.
-pub struct ConsensusManager {
-    /// The emission schedule to use for coinbase rewards
-    emission_schedule: EmissionSchedule,
+#[derive(Debug, Error, Clone, PartialEq)]
+pub enum ConsensusManagerError {
+    /// Difficulty adjustment encountered an error
+    DifficultyAdjustmentError(DifficultyAdjustmentError),
+    /// Difficulty adjustment manager encountered an error
+    DifficultyAdjustmentManagerError(DiffAdjManagerError),
+    /// Problem with the DB backend storage
+    ChainStorageError(ChainStorageError),
+    /// There is no blockchain to query
+    EmptyBlockchain,
+    /// RwLock access broken
+    PoisonedAccess,
+    /// No Diffuclty adjustment manager present
+    MissingDifficultyAdjustmentManager,
 }
 
-impl ConsensusManager {
-    pub fn new() -> Self {
-        ConsensusManager::default()
-    }
+/// This is the consensus manager struct. This manages all state-full consensus code.
+/// The inside is wrapped inside of an ARC so that it can safely and cheaply be cloned.
+/// The code is multi-thread safe and so only one instance is required. Inner objects are wrapped inside of RwLocks.
+pub struct ConsensusManager<B>
+where B: BlockchainBackend
+{
+    inner: Arc<ConsensusManagerInner<B>>,
+}
 
+impl<B> ConsensusManager<B>
+where B: BlockchainBackend
+{
+    /// Get a pointer to the emission schedule
     pub fn emission_schedule(&self) -> &EmissionSchedule {
-        &self.emission_schedule
+        &self.inner.emission_schedule
+    }
+
+    /// This moves over a difficulty adjustment manager to the ConsensusManager to control.
+    pub fn set_diff_manager(&self, diff_manager: DiffAdjManager<B>) -> Result<(), ConsensusManagerError> {
+        let mut lock = self
+            .inner
+            .diff_adj_manager
+            .write()
+            .map_err(|_| ConsensusManagerError::PoisonedAccess)?;
+        *lock = Some(diff_manager);
+        Ok(())
+    }
+
+    /// This returns the difficulty adjustment manager back. This can safely be cloned as the Difficulty adjustment
+    /// manager wraps an ARC in side of it.
+    pub fn get_diff_manager(&self) -> Result<DiffAdjManager<B>, ConsensusManagerError> {
+        match self
+            .inner
+            .diff_adj_manager
+            .read()
+            .map_err(|_| ConsensusManagerError::PoisonedAccess)?
+            .as_ref()
+        {
+            Some(v) => Ok(v.clone()),
+            None => Err(ConsensusManagerError::MissingDifficultyAdjustmentManager),
+        }
```

Comment:
This can be pulled out into a helper function and elimate all the boilerplate. See for example `access_metadata`
in blockchain_database.rs#276